### ISSUE TITLE
# Dot product between two 2d NArray matrices, one single-row and one single-column, is not correct. A strange warning is printed though no Error is raised.

### ISSUE THREAD

**Issue by giuse** [ruby-numo/numo-narray#86]
Dot product between two 2d NArray matrices, one single-row and one single-column, is not correct. A strange warning is printed though no Error is raised.

```ruby
a = Numo::DFloat.new(1,3).seq # => Numo::DFloat#shape=[1,3]  [[0, 1, 2]]
b = Numo::DFloat.new(3,1).seq # => Numo::DFloat#shape=[3,1]  [[0], [1], [2]]
a.dot(b)
# => ** On entry to DGEMM  parameter number 13 had an illegal value
# => Numo::DFloat#shape=[1,1]  [[0]]
# Expected: Numo::DFloat#shape=[1,1] [[5]]
```

**Comment by Try2Code** [comment]
the built-in tests cover "dot". do you get the same error running them?

**Comment by masa16** [comment]
If numo-linalg is loaded, NArray#dot uses Linalg.dot method.
Otherwise it uses NArray#mulsum method.
This seems a bug in Blas.dgemm.

**Comment by Try2Code** [comment]
I dont get the error on my system (archlinux, ruby-2.5, numo-narray (226e3ee), numo-linalg (c0ac1ad)

**Comment by giuse** [comment]
I conferm numo-linalg is loaded. Did not consider that, apologies. Here's an updated example tested on a clean console:

```ruby
require 'numo/narray'
# => true
a = Numo::DFloat.new(1,3).seq
# => Numo::DFloat#shape=[1,3]  [[0, 1, 2]]
b = Numo::DFloat.new(3,1).seq
# => Numo::DFloat#shape=[3,1]  [[0], [1], [2]]
a.dot(b)
# => Numo::DFloat#shape=[1,1]  [[5]]
require 'numo/linalg'
# => true
a.dot(b)
# ** On entry to DGEMM  parameter number 13 had an illegal value
# => Numo::DFloat#shape=[1,1]  [[0]]
```

**Comment by Try2Code** [comment]
I get a load error

```% ruby testDot.rb                                                               [Fri 2018-03-23|19:32:40]
Numo::DFloat#shape=[1,1]
[[5]]
Traceback (most recent call last):
	5: from testDot.rb:9:in `<main>'
	4: from /usr/lib/ruby/site_ruby/2.5.0/rubygems/core_ext/kernel_require.rb:39:in `require'
	3: from /usr/lib/ruby/site_ruby/2.5.0/rubygems/core_ext/kernel_require.rb:135:in `rescue in require'
	2: from /usr/lib/ruby/site_ruby/2.5.0/rubygems/core_ext/kernel_require.rb:135:in `require'
	1: from /home/ram/.gem/ruby/2.5.0/gems/numo-linalg-0.1.1/lib/numo/linalg.rb:3:in `<top (required)>'
/home/ram/.gem/ruby/2.5.0/gems/numo-linalg-0.1.1/lib/numo/linalg/loader.rb:176:in `load_library': cannot find backend library for Numo::Linalg (RuntimeError)
```
when running 
```% cat testDot.rb                                                                [Fri 2018-03-23|19:34:30]
require 'numo/narray'

a = Numo::DFloat.new(1,3).seq

b = Numo::DFloat.new(3,1).seq

pp a.dot(b)

require 'numo/linalg'

pp a.dot(b)
```

ruby 2.5.0p0 (2017-12-25 revision 61468)

**Comment by Try2Code** [comment]
I fixed that - I had a installed a blas library, but linalg needs lapack. with C-interface to lapack, I get the message @giuse reported

**Comment by giuse** [comment]
> If numo-linalg is loaded, NArray#dot uses Linalg.dot method.
> Otherwise it uses NArray#mulsum method.
> This seems a bug in Blas.dgemm.

A Google search for Blad.dgemm yielded multiple correspondences. Where can we correctly direct/file the issue? Also, would switching library avoid this? I am currently on OpenBLAS.

In the meanwhile I did this:

```ruby
class Numo::DFloat
  alias dot mulsum
end
```

Which is clearly not optimal since `#dot` is defined on `Numo::NArray` while `#mulsum` is defined on its typed subclasses (hence my reference to `Numo::DFloat`).
Any advice?

**Comment by giuse** [comment]
**Update:** the alias above is clearly not correct.
```ruby
require 'numo/narray'
# => true
a = Numo::DFloat.new(1,3).seq
# => Numo::DFloat#shape=[1,3]  [[0, 1, 2]]
b = Numo::DFloat.new(3,1).seq
# => Numo::DFloat#shape=[3,1]  [[0], [1], [2]]
a.dot(b)
# => Numo::DFloat#shape=[1,1]  [[5]]
a.mulsum(b)
# => 9.0
```
Can the original `Numo::NArray#dot` method be accessed after `require 'numo/linalg'`?
I guess I will alias it out before loading Linalg and setting it back after for now.

**Comment by masa16** [comment]
I have identified the reason of dgemm bug and it will be fixed.

NArray#dot is defined as:
https://github.com/ruby-numo/numo-narray/blob/master/lib/numo/narray/extra.rb#L1081-L1116

NArray#mulsum is inferior to BLAS implementation, so it should be replaced if BLAS is available.

**Comment by masa16** [comment]
Fixed a bug in numo-narray and released ver. 0.9.1.2.

**Comment by giuse** [comment]
I confirm this is now fixed in 0.9.1.2
```ruby
require 'numo/narray'
# => true
a = Numo::DFloat.new(1,3).seq
# => Numo::DFloat#shape=[1,3]  [[0, 1, 2]]
b = Numo::DFloat.new(3,1).seq
# => Numo::DFloat#shape=[3,1]  [[0], [1], [2]]
a.dot(b)
# => Numo::DFloat#shape=[1,1]  [[5]]
require 'numo/linalg'
# => true
a.dot(b)
# => Numo::DFloat#shape=[1,1]  [[5]]
Numo::Linalg.dot(a,b)
# => Numo::DFloat#shape=[1,1]  [[5]]
```
Thanks a lot!
